### ISSUE TITLE
Performance Discrepancy

### ISSUE THREAD

**Issue by micycle1** [gwlucastrig/Tinfour#84]
In the docs, Tin processing time scales linearly and processes in the order of one million points per second:
![image](https://user-images.githubusercontent.com/9304234/166112122-0bbda304-ec49-484a-aa59-343029aa94d3.png)

Yet, the following code takes ~125s, and playing around with different values for N, performance is very much non-linear.

```
int N = 1_000_000;
List<Vertex> vertices = new ArrayList<>();
for (int i = 0; i < N; i++) {
	double x = Math.random() * 1000;
	double y = Math.random() * 1000;
	vertices.add(new Vertex(x, y, 0));
}
long tStart = 0;
IIncrementalTin tin = new IncrementalTin();
tStart = System.nanoTime();
tin.add(vertices, null);
long tTinfour = System.nanoTime() - tStart;
```

10,000 points: ~65ms
100,000 points: ~2.7s
200,000 points: ~9.4s

What's going on?

**Comment by micycle1** [comment]
Ahah, I found the hilbert sort shortly afterwards...
Processing the vertices, having sorted them first with this, takes the time to process 1 million points down to <1s.
```
HilbertSort hs = new HilbertSort();
hs.sort(vertices);
```

It might be worth increasing the visibility of this / mentioning the need to do this in docs, or some kind of method/constructor for Tin construction from (random) point collections.

**Comment by gwlucastrig** [comment]
Yes. You've found the worst-case data sample for Tinfour performance. I'm not really sure what to do about this.

For random data, I believe that the time complexity of an insertion is roughly N to the power 3/2, but I haven't done a rigorous analysis and it could be worse.

The limiting factor on the insertion algorithm is the time it takes to find out what triangle contains the point to be inserted. Ordinarily, the insertion logic takes advantage of the idea that, for many data sets,  "a point to be inserted next is usually close to the point that was inserted last".   In the case of random data, this assumption is not true at all.

Here's something you might find interesting.  If at the end of your insertion process, you call:

     tin.printDiagnostics(System.out);

the class prints out a set of diagnostics.  If you look at the construction statistics for your non-sorted data you see:
 
    Construction statistics
    Number of SLW walks:           1000000
       exterior phase:                 234
       tests:                     1032616994
       extended:                         6
       avg steps to completion:        801.84
    InCircle calculations:         8998441
 
Note the large number of tests and avg steps to completion.  Now look at the statistics if the sort is applied.

    Construction statistics
    Number of SLW walks:           1000000
       exterior phase:                8996
       tests:                      6137206
       extended:                         6
       avg steps to completion:          3.13
    InCircle calculations:        12168787

The key difference is the length of the Stochastic Lawson Walks (SLW).  801.84 steps for the unsorted data version versus about 3.13 for the sorted.  For random data, the number of steps for each insertion averages about SQRT(N).   For the sorted data, it's close to a constant number of 3 steps or so.

Some of this is explained in the PDF document [Data Elements and Algorithms for the Tinfour Library](https://gwlucastrig.github.io/TinfourDocs/TinfourAlgorithmsAndDataElements.pdf)

**Comment by gwlucastrig** [comment]
Incidentally, I was looking through my Tinfour test project in Netbeans to pick a good place to store your test program, and I realized I have a package called test.micycle1.  I think that's a good tribute. Thanks for sending me so many interesting questions over the last few years.

Gary

**Comment by gwlucastrig** [comment]
Michael,

I think that you are right that I have to make the recommendation for using the Hilbert sort much more prominent.  It's hard for me to look at this stuff with a fresh perspective, so I was hoping you might have a suggestion.  Is there a place in the documentation that I could include the discussion so that it would have the highest probability of a new user finding it?

Also, I did a test where I processed a random number of samples between 1000 and 1 million with samples at random coordinates. It turns out that the best fit for the time-complexity curve for the non-sorted sample set was of O(N^2).  The Hilbert sort reduces that to basically O(N) with a small quadratic component.

Non-sorted:

![ProcessingRateRandomSamples](https://user-images.githubusercontent.com/19698145/166342752-f0e604ce-d563-4bbb-87a4-d3c240818b72.png)

But, sorted, with a much smaller processing time:
![ProcessingRateRandomSamplesSorted](https://user-images.githubusercontent.com/19698145/166342847-0a94a3a1-6b48-469c-9ba2-d201d418c1d2.png)

I've attached a zip file showing the test program.  The other thing I would probably try is to include the time-to-sort as part of the processing-time tabulation.  Properly speaking, it does contribute to the overall processing time.

[TimeComplexity.zip](https://github.com/gwlucastrig/Tinfour/files/8606585/TimeComplexity.zip)
